pyproject.toml

```diff
@@ -1,10 +1,259 @@
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+]
+dependencies = [
+    "zope.interface >= 4.4.2",
+    "constantly >= 15.1",
+    "incremental >= 21.3.0",
+    "Automat >= 0.8.0",
+    "hyperlink >= 17.1.1",
+    "attrs >= 19.2.0",
+    "typing_extensions >= 3.6.5",
+    "twisted-iocpsupport >= 1.0.2, <2; platform_system == 'Windows'",
+]
+
+[project.optional-dependencies]
+# This is list of minimum dependencies required to run the tests.
+# The plan is to remove the `test` extra requirement and skip tests that
+# depend on them
+# To prevent trunk failures we should pin all dev dependencies to minor
+# versions.
+# We trust semantic versioning and auto-upgrading to a bugfix release
+# should be OK.
+test = [
+    "cython-test-exception-raiser >= 1.0.2, <2",
+    "PyHamcrest >= 1.9.0",
+    "hypothesis ~= 6.0",
+]
+
+# List of dependencies required to build the documentation and test the
+# release scripts and process.
+dev-release = [
+    "towncrier ~= 22.8",
+    "pydoctor ~= 22.9.0",
+    "sphinx-rtd-theme ~= 0.5",
+    "readthedocs-sphinx-ext ~= 2.1",
+    "sphinx >= 4.1.2, <6",
+]
+dev_release = [
+    "twisted[dev-release]",
+]
+
+# All the extra tools used to help with the development process.
+dev = [
+    "twisted[dev-release]",
+    "pyflakes ~= 2.2",
+    # TODO: support python-subunit in py3.10 https://twistedmatrix.com/trac/ticket/10115
+    "python-subunit ~= 1.4; python_version < '3.10'",
+    "twistedchecker ~= 0.7",
+    "coverage >= 6b1, <7",
+]
+
+tls = [
+    "pyopenssl >= 21.0.0",
+    # service_identity 18.1.0 added support for validating IP addresses in
+    # certificate subjectAltNames
+    "service_identity >= 18.1.0",
+    "idna >= 2.4",
+]
+
+conch = [
+    "pyasn1",
+    "cryptography >= 2.6",
+    "appdirs >= 1.4.0",
+    "bcrypt >= 3.0.0",
+]
+
+conch-nacl = [
+    "twisted[conch]",
+    # Used to support Ed25519 keys on systems with OpenSSL < 1.1.1b
+    "PyNaCl",
+]
+conch_nacl = [
+    "twisted[conch-nacl]",
+]
+
+serial = [
+    "pyserial >= 3.0",
+    "pywin32 != 226; platform_system == 'Windows'",
+]
+
+http2 = [
+    "h2 >= 3.0, < 5.0",
+    "priority >= 1.1.0, < 2.0",
+]
+
+contextvars = [
+    "contextvars >= 2.4, < 3; python_version < '3.7'",
+]
+
+all-non-platform = [
+    "twisted[test,tls,conch,serial,http2,contextvars]",
+]
+all_non_platform = [
+    "twisted[all-non-platform]",
+]
+
+macos-platform = [
+    "twisted[all-non-platform]",
+    "pyobjc-core",
+    "pyobjc-framework-CFNetwork",
+    "pyobjc-framework-Cocoa",
+]
+macos_platform = [
+    "twisted[macos-platform]",
+]
+
+windows-platform = [
+    "twisted[all-non-platform]",
+    "pywin32 != 226",
+]
+windows_platform = [
+    "twisted[windows-platform]",
+]
+
+osx-platform = [
+    "twisted[macos-platform]",
+]
+osx_platform = [
+    "twisted[osx-platform]",
+]
+
+gtk-platform = [
+    "twisted[all-non-platform]",
+    "pygobject",
+]
+gtk_platform = [
```

Comment:
Can we have all these backward compatible definitions in a single place and with a comment, to explain why we have them and how to maintain them?

```suggestion
# We define all the legacy dependency targets using the
# underscore for backward compatibility.
# Once underscores are no longer supported by Python package tools
# like pip and PyPI, these can be removed.
# Any new extra dependency should not be added here.
# This list is here only for backward compatiblility.
dev_release = ["twisted[dev-release]"]
conch_nacl = ["twisted[conch-nacl]"]
gtk_platform = [
```